src/core/FaceControls.tsx

```diff
@@ -0,0 +1,413 @@
+          },
+          pause: () => {
+            webcamApiRef.current?.videoTextureApiRef.current?.texture.source.data.pause()
+          },
+        }),
+      [detect, computeTarget, update]
+    )
+    useImperativeHandle(fref, () => api, [api])
+
+    //
+    // events callbacks
+    //
+
+    useEffect(() => {
+      const onVideoFrameCb = (e: THREE.Event) => {
+        if (!manualDetect) detect(e.texture.source.data, e.time)
+        if (onVideoFrame) onVideoFrame(e)
+      }
+
+      api.addEventListener('videoFrame', onVideoFrameCb)
+
+      return () => {
+        api.removeEventListener('videoFrame', onVideoFrameCb)
+      }
+    }, [api, detect, faceLandmarker, manualDetect, onVideoFrame])
+
+    // `controls` global state
+    useEffect(() => {
+      if (makeDefault) {
+        const old = get().controls
+        set({ controls: api })
+        return () => set({ controls: old })
+      }
+    }, [makeDefault, api, get, set])
+
+    const points = faces?.faceLandmarks[0]
+    const facialTransformationMatrix = faces?.facialTransformationMatrixes?.[0]
+    const faceBlendshapes = faces?.faceBlendshapes?.[0]
+    return (
+      <FaceControlsContext.Provider value={api}>
+        {webcam && <Webcam ref={webcamApiRef} autostart={autostart} videoTextureSrc={webcamVideoTextureSrc} />}
+
+        <Facemesh
+          ref={facemeshApiRef}
+          {...facemesh}
+          points={points}
+          depth={depth}
+          facialTransformationMatrix={facialTransformationMatrix}
+          faceBlendshapes={faceBlendshapes}
+          eyes={eyes}
+          eyesAsOrigin={eyesAsOrigin}
+          offset={offset}
+          offsetScalar={offsetScalar}
+          debug={debug}
+          rotation-z={Math.PI}
+          visible={debug}
+        >
+          <meshStandardMaterial flatShading={true} side={THREE.DoubleSide} />
+        </Facemesh>
+      </FaceControlsContext.Provider>
+    )
+  }
+)
+
+export const useFaceControls = () => useContext(FaceControlsContext)
+
+//
+// Webcam
+//
+
+type WebcamApi = {
+  videoTextureApiRef: RefObject<VideoTextureApi>
+}
+
+type WebcamProps = {
+  videoTextureSrc?: VideoTextureSrc
+  autostart?: boolean
+}
+
+const Webcam = forwardRef<WebcamApi, WebcamProps>(({ videoTextureSrc, autostart = true }, fref) => {
+  const [stream, setStream] = useState<MediaStream>()
+
+  const videoTextureApiRef = useRef<VideoTextureApi>(null)
+
+  const faceControls = useFaceControls()
+  useEffect(() => {
+    let strm: MediaStream
+
+    if (!videoTextureSrc) {
+      navigator.mediaDevices
+        .getUserMedia({
+          audio: false,
+          video: { facingMode: 'user' },
+        })
+        .then((s) => {
+          strm = s
+          faceControls.dispatchEvent({ type: 'stream', stream: strm })
+          setStream(strm)
+        })
+        .catch(console.error)
+    }
+
+    return () => strm?.getTracks().forEach((track) => track.stop())
+  }, [faceControls, videoTextureSrc])
```

Comment:
Suspense would also be helpful here. I'm pedantic about this now as it's breaking later.

Reply:
ok now done ;) I let you check my impl